oak-run/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldVersionUtils.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.indexversion;
+
+import org.apache.jackrabbit.oak.api.CommitFailedException;
+import org.apache.jackrabbit.oak.commons.PathUtils;
+import org.apache.jackrabbit.oak.plugins.index.lucene.property.RecursiveDelete;
+import org.apache.jackrabbit.oak.plugins.index.search.spi.query.IndexName;
+import org.apache.jackrabbit.oak.spi.commit.CommitInfo;
+import org.apache.jackrabbit.oak.spi.commit.EmptyHook;
+import org.apache.jackrabbit.oak.spi.state.NodeBuilder;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+import org.apache.jackrabbit.oak.spi.state.NodeStateUtils;
+import org.apache.jackrabbit.oak.spi.state.NodeStore;
+import org.apache.jackrabbit.util.ISO8601;
+import org.jetbrains.annotations.NotNull;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class PurgeOldVersionUtils {
+
+    public static final String OAK_INDEX = "oak:index";
+
+    public static long getMillisFromString(String strDate) {
+        long millis = ISO8601.parse(strDate).getTimeInMillis();
+        return millis;
+    }
+
+    /**
+     * @param nodeBuilder
+     * @param path        Path of node whose nodeBuilder object should be returned.
+     * @return nodeBuilder object of node at @param{path}
+     */
+    public static NodeBuilder getNode(@NotNull NodeBuilder nodeBuilder, @NotNull String path) {
+        for (String name : PathUtils.elements(checkNotNull(path))) {
+            nodeBuilder = nodeBuilder.getChildNode(checkNotNull(name));
```

Comment:
I tend to not re-assign the method's parameters. You could create a temp variable here.

On a separate note, this type of logic refers to a [**fold**](https://en.wikipedia.org/wiki/Fold_(higher-order_function)) in the functional paradigm. This usually helps to write this logic in a more idiomatic way. In java this is possible but it requires a bit more "ceremony" compared to other languages:

```java
return StreamSupport.stream(PathUtils.elements(checkNotNull(path)).spliterator(), false)
    .reduce(nodeBuilder,
        (nb, name) -> nb.getChildNode(checkNotNull(name)),
        (nb, nb2) -> nb); // combinator function, not need in sequential reductions but in java needs to be specified :-(
```

Here you can find a nice explanation of what happens underneath https://stackoverflow.com/a/33971436/1026224

Reply:
In my view, it's a matter of preference whether to use the functional style or the procedural. I personally prefer the procedural style, at least here. And here, I don't think re-assigning the parameter is a problem. I would say, do what you prefer.